src/backend/tcop/postgres.c

```diff
@@ -1560,16 +1561,23 @@ CheckDebugDtmActionSqlCommandTag(const char *sqlCommandTag)
 }
 
 static void
-restore_guc_to_QE(void )
+send_guc_to_QE(List *guc_list, bool is_restore)
 {
-	Assert(Gp_role == GP_ROLE_DISPATCH && gp_guc_restore_list);
+	Assert(Gp_role == GP_ROLE_DISPATCH && guc_list);
 	ListCell *lc;
+	MemoryContext oldcontext = CurrentMemoryContext;
 
 	start_xact_command();
 
-	foreach(lc, gp_guc_restore_list)
+	foreach(lc, guc_list)
 	{
 		struct config_generic* gconfig = (struct config_generic *)lfirst(lc);
+
+		/* We can't SET GUCs whose context is too high. */
+		if (gconfig->context < PGC_SUSET ||
+			(!is_restore && !(gconfig->flags & GUC_GPDB_NEED_SYNC)))
```

Comment:
Maybe this condition will be more understandable?
```suggestion
		if (!((gconfig->context >= PGC_SUSET) && (is_restore || (gconfig->flags & GUC_GPDB_NEED_SYNC)))
```

Reply:
Or maybe this would be right?
```suggestion
		if (!(is_restore || ((gconfig->context >= PGC_SUSET) && (gconfig->flags & GUC_GPDB_NEED_SYNC)))
```

Reply:
I think unchanged condition is simpler
`"If the context is too high || (this is not restore && we don't have a sync flag set)"`

Reply:
I think `PostgresMain` should send all GUCs, not only if `gconfig->context >= PGC_SUSET`, as it was before patch.

Reply:
It's impossible, since SET does not allow changing GUCs with context higher than PGC_SUSET, and we can't distinguish SET commands between sent by this function or users performing a manual SET. Before this patch, this was still the case.

Reply:
Before patch `PostgresMain` calls `restore_guc_to_QE` and it restores all GUCs in `gp_guc_restore_list`.
After patch `PostgresMain` calls `restore_guc_to_QE` with true as `is_restore` and now it does not restore GUCS `from gp_guc_restore_list` if they `context < PGC_SUSET`.

Reply:
SET command calls `set_config_option()` with `PGC_SUSET`/`PGC_USERSET`:
https://github.com/arenadata/gpdb/blob/cdc9332854a6aca15bfe98541a976342658be028/src/backend/utils/misc/guc.c#L7278-L7284
And then the context of the GUC is compared with the one passed as an argument:
https://github.com/arenadata/gpdb/blob/cdc9332854a6aca15bfe98541a976342658be028/src/backend/utils/misc/guc.c#L5928

So even if I get rid of this check, SET will not be able to restore any GUCs if the context is too high.

Reply:
Why not exclude continue when is_restore is true?